exec/java-exec/src/main/java/org/apache/drill/exec/store/PluginBootstrapLoaderImpl.java

```diff
@@ -0,0 +1,298 @@
+ * unreliable system: an upgrade file exists. Any upgrades are applied on top
+ * of the user's stored values. The upgrade file is then deleted to avoid
+ * applying the same upgrades multiple times. Not idea, but it is what it
+ * is for now.
+ *
+ * <h4>Format Plugins</h4>
+ *
+ * Earlier versions of this code provided support for format plugins. However,
+ * this code never worked. The original design of Drill has format plugins
+ * as a set of attributes of the file system plugin config. Additional work
+ * is needed to allow file system configs to share plugins. Such work is a
+ * good idea, so the ill-fated format plugin code still appears, but does
+ * nothing.
+ *
+ * <h4>Config Properties</h4>
+ *
+ * <dl>
+ * <dt>{@code ExecConstants.BOOTSTRAP_STORAGE_PLUGINS_FILE}
+ * ({@code drill.exec.storage.bootstrap.storage})</dt>
+ * <dd>The name of the bootstrap file, normally
+ * {@code bootstrap-storage-plugins.json}.</dd>
+ * <dt>{@code ExecConstants.UPGRADE_STORAGE_PLUGINS_FILE}
+ * ({@code drill.exec.storage.upgrade.storage})</dt>
+ * <dd>The name of the upgrade file, normally
+ * {@code storage-plugins-override.conf}. Unlike th bootstrap
+ * file, only one upgrade file can appear in the class path.
+ * The upgrade file is optional.</dd>
+ * </dl>
+ */
+public class PluginBootstrapLoaderImpl implements PluginBootstrapLoader {
+  private static final Logger logger = LoggerFactory.getLogger(PluginBootstrapLoaderImpl.class);
+
+  private final PluginRegistryContext context;
+
+  /**
+   * The file read the first time a Drillbit starts up, and deleted
+   * afterwards. A poor-man's way of handling upgrades. Will be
+   * non-null when an upgrade is needed, null thereafter until a
+   * new release.
+   */
+  private URL pluginsOverrideFileUrl;
+
+  public PluginBootstrapLoaderImpl(PluginRegistryContext context) {
+    this.context = context;
+  }
+
+  @Override
+  public StoragePlugins bootstrapPlugins() throws IOException {
+    StoragePlugins bootstrapPlugins = loadBootstrapPlugins();
+
+    // Upgrade the bootstrap plugins with any updates. Seems odd,
+    // but this is how Drill 1.17 works, so keeping the code.
+    // Uses the enabled status from the updates if different than
+    // the bootstrap version.
+
+    StoragePlugins updatedPlugins = updatedPlugins();
+    if (updatedPlugins != null) {
+      bootstrapPlugins.putAll(updatedPlugins);
+    }
+    return bootstrapPlugins;
+  }
+
+  @VisibleForTesting
+  protected StoragePlugins loadBootstrapPlugins() throws IOException {
+    // bootstrap load the config since no plugins are stored.
+    logger.info("No storage plugin instances configured in persistent store, loading bootstrap configuration.");
+    String storageBootstrapFileName = context.config().getString(ExecConstants.BOOTSTRAP_STORAGE_PLUGINS_FILE);
+    Set<URL> storageUrls = ClassPathScanner.forResource(storageBootstrapFileName, false);
+    // Format plugins does not work, see note below.
```

Comment:
How come this does not work? It was testes and did work.
This was added (see https://issues.apache.org/jira/browse/DRILL-7030) to allow loading formats from contrib module. This definitely should work. All formats in contrib folder contain `bootstrap-format-plugins.json` file. If you comment out this code, original behavior will be broken. Please make sure it works. Thanks.

Reply:
Well, I am not sure we are on the same page. Let's me try to explain the intension of the written code.

First we have `bootstrap-storage-plugins.json` which resides in exec module and has a list of storage plugins used as default on fresh start up. Then we have contrib module where some plugins and formats reside. We want them to be included in default plugins / formats as well during fresh start up. All works fine for the plugins, since each plugin also has `bootstrap-storage-plugins.json` and during Drill start-up they are loaded. 

The problem was with formats which resided in contrib module. Since when new format was added, developer had to modify `bootstrap-storage-plugins.json` in exec module to ensure new format is included. At this point `bootstrap-format-plugins.json` was added. It allowed to include formats from contrib module as bootstrap. Yes, developer has to indicate plugin name, for instance, `dfs` where format should be added but this allowed to make formats fully pluggable. This was one of the really nice improvements in the previous release. So commenting this code means you are removing previous functionality.

Steps how to check that functionality works.
1. Build Drill.
2. Make sure previous build data was removed (ex: `/tmp/drill` folder).
3. Start Drill.
4. Open Web UI Storage tab.
5. Open `dfs` Storage plugin. 
6. Check that it contains all formats, including those from contrib module. For example, `ltsv`.

Reply:
Restored the format bootstrap functionality. At present, it will only install formats at bootstrap time, it won't add new formats to existing configs which, it seems, is how it worked in Drill 1.17. Also added a unit test.